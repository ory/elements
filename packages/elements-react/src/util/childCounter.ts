// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { Children, ReactNode, isValidElement } from "react"

export function countRenderableChildren(children: ReactNode | ReactNode[]) {
  return Children.toArray(children).filter((c) => {
    if (isValidElement(c)) {
      return true
    }
    return false
  }).length
}
