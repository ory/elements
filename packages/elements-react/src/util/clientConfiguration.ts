// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { ConfigurationParameters } from "@ory/client-fetch"
import { IntlContextProps } from "../context/intl-context"

export type IntlConfig = IntlContextProps

export type OryClientConfiguration = {
  /**
   * The name of the application the user is logging in to.
   */
  name: string

  /**
   * An optional logo URL to display in the UI instead of the name.
   */
  logoUrl?: string

  stylesheet?: string

  favicon?: string

  sdk: {
    url: string
    options?: Partial<ConfigurationParameters>
  }

  project: {
    registration_enabled: boolean
    verification_enabled: boolean
    recovery_enabled: boolean

    recovery_ui_url: string
    registration_ui_url: string
    verification_ui_url: string
    login_ui_url: string
  }
  intl?: IntlConfig
}
