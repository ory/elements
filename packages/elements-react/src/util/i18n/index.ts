// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { UiText } from "@ory/client-fetch"
import { IntlShape } from "react-intl"

/**
 * Converts a UiText to a FormattedMessage.
 * The UiText contains the id of the message and the context.
 * The context is used to inject values into the message from Ory, e.g. a timestamp.
 * For example a UI Node from Ory might look like this:
 *
 * \{
 *  "type":"input",
 *  "group":"default",
 *  "attributes": \{
 *      "name":"traits.email",
 *      "type":"email",
 *      "required":true,
 *      "autocomplete":"email",
 *      "disabled":false,
 *      "node_type":"input"
 *  \},
 *  "messages":[],
 *  "meta": \{
 *    "label": \{
 *      "id":1070002,
 *      "text":"E-Mail",
 *      "type":"info",
 *      "context":\{
 *        "title":"E-Mail"
 *      \},
 *    \}
 *  \}
 * \}
 *
 * The context has the key "title" which matches the formatter template name "\{title\}"
 * An example translation file would look like this:
 * \{
 *  "identities.messages.1070002": "\{title\}"
 * \}
 *
 * The formatter would then take the meta.label.id and look for the translation with the key matching the id.
 * It would then replace the template "\{title\}" with the value from the context with the key "title".
 *
 * @param uiText - The UiText is part of the UiNode object sent by Kratos when performing a flow.
 * @param intl - The intl object from react-intl
 */
export const uiTextToFormattedMessage = (
  { id, context = {}, text }: Omit<UiText, "type">,
  intl: IntlShape,
) => {
  const contextInjectedMessage = Object.entries(context).reduce(
    (accumulator, [key, value]) => {
      // context might provide an array of objects instead of a single object
      // for example when looking up a recovery code
      if (Array.isArray(value)) {
        return {
          ...accumulator,
          [key]: value,
          [key + "_list"]: intl.formatList<string>(value),
        }
      } else if (key.endsWith("_unix")) {
        if (typeof value === "number") {
          return {
            ...accumulator,
            [key]: intl.formatDate(new Date(value * 1000)),
            [key + "_since"]: intl.formatDateTimeRange(
              new Date(value),
              new Date(),
            ),
            [key + "_since_minutes"]: Math.abs(
              (value - new Date().getTime() / 1000) / 60,
            ).toFixed(2),
            [key + "_until"]: intl.formatDateTimeRange(
              new Date(),
              new Date(value),
            ),
            [key + "_until_minutes"]: Math.abs(
              (new Date().getTime() / 1000 - value) / 60,
            ).toFixed(2),
          }
        }
      }
      return {
        ...accumulator,
        [key]: value as string | number,
      }
    },
    {},
  )

  return intl.formatMessage(
    {
      id: `identities.messages.${id}`,
      defaultMessage: text,
    },
    contextInjectedMessage,
  )
}
