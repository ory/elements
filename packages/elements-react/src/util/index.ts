// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export * from "./clientConfiguration"
export * from "./flowContainer"
export * from "./i18n"
export * from "./submitHandler"
export * from "./test-id"
