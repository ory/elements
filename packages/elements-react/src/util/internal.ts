// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export function replaceWindowFlowId(flow: string) {
  const url = new URL(window.location.href)
  url.searchParams.set("flow", flow)
  window.location.href = url.toString()
}
