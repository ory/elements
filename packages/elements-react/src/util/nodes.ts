// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { UiNode, UiNodeInputAttributes } from "@ory/client-fetch"

export function findScreenSelectionButton(
  nodes: UiNode[],
): { attributes: UiNodeInputAttributes } | undefined {
  return nodes.find(
    (node) =>
      node.attributes.node_type === "input" &&
      node.attributes.type === "submit" &&
      node.attributes.name === "screen",
  ) as { attributes: UiNodeInputAttributes }
}
