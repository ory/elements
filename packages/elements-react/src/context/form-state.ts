// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { FlowType, UiNode, UiNodeGroupEnum } from "@ory/client-fetch"
import { useReducer } from "react"
import { isChoosingMethod } from "../components/card/card-two-step.utils"
import { OryFlowContainer } from "../util"
import { nodesToAuthMethodGroups } from "../util/ui"

export type FormState =
  | { current: "provide_identifier" }
  | { current: "select_method" }
  | { current: "method_active"; method: UiNodeGroupEnum }
  | { current: "success_screen" }
  | { current: "settings" }

export type FormStateAction =
  | {
      type: "action_flow_update"
      flow: OryFlowContainer
    }
  | {
      type: "action_select_method"
      method: UiNodeGroupEnum
    }

function findMethodWithMessage(nodes?: UiNode[]) {
  return nodes
    ?.filter((n) => !["default", "identifier_first"].includes(n.group))
    ?.find((node) => node.messages?.length > 0)
}

function parseStateFromFlow(flow: OryFlowContainer): FormState {
  switch (flow.flowType) {
    case FlowType.Registration:
    case FlowType.Login: {
      const methodWithMessage = findMethodWithMessage(flow.flow.ui.nodes)
      if (flow.flow.active == "link_recovery") {
        return { current: "method_active", method: "link" }
      } else if (flow.flow.active == "code_recovery") {
        return { current: "method_active", method: "code" }
      } else if (methodWithMessage) {
        return { current: "method_active", method: methodWithMessage.group }
      } else if (
        flow.flow.active &&
        !["default", "identifier_first", "oidc"].includes(flow.flow.active)
      ) {
        return { current: "method_active", method: flow.flow.active }
      } else if (isChoosingMethod(flow.flow.ui.nodes)) {
        // Login has a special case where we only have one method. Here, we
        // do not want to display the chooser.
        const authMethods = nodesToAuthMethodGroups(flow.flow.ui.nodes)
        if (authMethods.length === 1) {
          return { current: "method_active", method: authMethods[0] }
        }
        return { current: "select_method" }
      } else if (flow.flow.ui.messages?.some((m) => m.id === 1010016)) {
        // Account linking edge case
        return { current: "select_method" }
      }
      return { current: "provide_identifier" }
    }
    case FlowType.Recovery:
    case FlowType.Verification:
      // The API does not provide types for the active field of the recovery flow
      // TODO: Add types for the recovery flow in Kratos
      if (flow.flow.active === "code" || flow.flow.active === "link") {
        if (flow.flow.state === "choose_method") {
          return { current: "provide_identifier" }
        }
        return { current: "method_active", method: flow.flow.active }
      }
      break
    case FlowType.Settings:
      return { current: "settings" }
  }
  console.warn(
    `[Ory/Elements React] Encountered an unknown form state on ${flow.flowType} flow with ID ${flow.flow.id}`,
  )
  throw new Error("Unknown form state")
}

export function formStateReducer(
  state: FormState,
  action: FormStateAction,
): FormState {
  switch (action.type) {
    case "action_flow_update":
      return parseStateFromFlow(action.flow)
    case "action_select_method":
      return { current: "method_active", method: action.method }
  }
  return state
}

export function useFormStateReducer(flow: OryFlowContainer) {
  return useReducer(formStateReducer, parseStateFromFlow(flow))
}
