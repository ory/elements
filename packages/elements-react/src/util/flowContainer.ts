// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  FlowError,
  FlowType,
  LoginFlow,
  RecoveryFlow,
  RegistrationFlow,
  SettingsFlow,
  VerificationFlow,
} from "@ory/client-fetch"
import { OryClientConfiguration } from "./clientConfiguration"

/**
 * A generic flow container, containing a flowType, the flow itself and the config object
 *
 * @see OryClientConfiguration
 */
type OryFlow<TFlowType, TFlow> = {
  flowType: TFlowType
  flow: TFlow
  config: OryClientConfiguration
}

/**
 * A flow container for the login flow
 *
 * @see OryFlow
 * @see LoginFlow
 */
export type LoginFlowContainer = OryFlow<FlowType.Login, LoginFlow>

/**
 * A flow container for the registration flow
 *
 * @see OryFlow
 * @see RegistrationFlow
 */
export type RegistrationFlowContainer = OryFlow<
  FlowType.Registration,
  RegistrationFlow
>

/**
 * A flow container for the recovery flow
 *
 * @see OryFlow
 * @see RecoveryFlow
 */
export type RecoveryFlowContainer = OryFlow<FlowType.Recovery, RecoveryFlow>

/**
 * A flow container for the verification flow
 *
 * @see OryFlow
 * @see VerificationFlow
 *
 */
export type VerificationFlowContainer = OryFlow<
  FlowType.Verification,
  VerificationFlow
>
/**
 * A flow container for the settings flow
 *
 * @see OryFlow
 * @see SettingsFlow
 */
export type SettingsFlowContainer = OryFlow<FlowType.Settings, SettingsFlow>

/**
 * A flow container for the error flow
 *
 * @see OryFlow
 * @see FlowError (Error flow)
 *
 */
export type ErrorFlowContainer = OryFlow<FlowType.Error, FlowError>

/**
 * A union type of all flow containers
 */
export type OryFlowContainer =
  | LoginFlowContainer
  | RegistrationFlowContainer
  | RecoveryFlowContainer
  | VerificationFlowContainer
  | SettingsFlowContainer
// TODO: Add ErrorFlowContainer
