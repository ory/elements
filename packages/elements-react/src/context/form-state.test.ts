// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { FlowType, UiNodeGroupEnum } from "@ory/client-fetch"
import { act, renderHook } from "@testing-library/react"
import { OryFlowContainer } from "../util"
import { useFormStateReducer } from "./form-state" // Adjust path as needed

const init = {
  flowType: FlowType.Login,
  flow: { ui: { nodes: [] } },
} as unknown as OryFlowContainer
test('should initialize with "provide_identifier" state', () => {
  const { result } = renderHook(() => useFormStateReducer(init))

  const [state] = result.current
  expect(state).toEqual({ current: "provide_identifier" })
})

test('should initialize with "settings" state for settings flows', () => {
  const { result } = renderHook(() =>
    useFormStateReducer({
      flowType: FlowType.Settings,
    } as unknown as OryFlowContainer),
  )

  const [state] = result.current
  expect(state).toEqual({ current: "settings" })
})

test('should transition to "method_active" state when "action_select_method" is dispatched', () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  act(() => {
    dispatch({
      type: "action_select_method",
      method: UiNodeGroupEnum.Code,
    })
  })

  const [state] = result.current
  expect(state).toEqual({
    current: "method_active",
    method: UiNodeGroupEnum.Code,
  })
})

const activeMethods = ["link_recovery", "code_recovery", "code"]

activeMethods.forEach((activeMethod) => {
  test(`should parse state from flow on "action_flow_update" with active method = ${activeMethod}`, () => {
    const { result } = renderHook(() => useFormStateReducer(init))
    const [, dispatch] = result.current

    const mockFlow = {
      flowType: FlowType.Login,
      flow: {
        active: activeMethod,
        ui: { nodes: [] }, // Assuming nodes structure
      },
    } as unknown as OryFlowContainer

    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    })

    const [state] = result.current
    expect(state).toMatchSnapshot()
  })
})

test(`should parse state from flow on "action_flow_update" provide_identifier`, () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Login,
    flow: {
      active: "",
      ui: { nodes: [] }, // Assuming nodes structure
    },
  } as unknown as OryFlowContainer

  act(() => {
    dispatch({
      type: "action_flow_update",
      flow: mockFlow,
    })
  })

  const [state] = result.current
  expect(state).toEqual({ current: "provide_identifier" })
})

test(`should parse state from flow on "action_flow_update" if password has validation error`, () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Login,
    flow: {
      active: "",
      ui: {
        nodes: [
          {
            attributes: {
              autocomplete: "new-password",
              disabled: false,
              name: "password",
              node_type: "input",
              required: true,
              type: "password",
            },
            group: "password",
            messages: [
              {
                context: {
                  actual_length: 2,
                  min_length: 8,
                },
                id: 4000032,
                text: "The password must be at least 8 characters long, but got 2.",
                type: "error",
              },
            ],
            meta: {
              label: {
                id: 1070001,
                text: "Password",
                type: "info",
              },
            },
            type: "input",
          },
        ],
      }, // Assuming nodes structure
    },
  } as unknown as OryFlowContainer

  act(() => {
    dispatch({
      type: "action_flow_update",
      flow: mockFlow,
    })
  })

  const [state] = result.current
  expect(state).toEqual({ current: "method_active", method: "password" })
})

test(`should ignore validation message on default group nodes`, () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Login,
    flow: {
      active: "",
      ui: {
        nodes: [
          {
            attributes: {
              autocomplete: "email",
              disabled: false,
              name: "traits.email",
              node_type: "input",
              required: true,
              type: "email",
              value: "fdghj",
            },
            group: "default",
            messages: [
              {
                context: {
                  reason: '"fdghj" is not valid "email"',
                },
                id: 4000001,
                text: '"fdghj" is not valid "email"',
                type: "error",
              },
            ],
            meta: {
              label: {
                context: {
                  title: "E-Mail",
                },
                id: 1070002,
                text: "E-Mail",
                type: "info",
              },
            },
            type: "input",
          },
        ],
      }, // Assuming nodes structure
    },
  } as unknown as OryFlowContainer

  act(() => {
    dispatch({
      type: "action_flow_update",
      flow: mockFlow,
    })
  })

  const [state] = result.current
  expect(state).toEqual({ current: "provide_identifier" })
})

test(`should parse state from flow on "action_flow_update" when choosing method on registration`, () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Registration,
    flow: {
      active: "",
      ui: {
        nodes: [
          {
            attributes: {
              name: "screen",
              value: "previous",
            },
          },
        ],
      }, // Assuming nodes structure
    },
  } as unknown as OryFlowContainer

  act(() => {
    dispatch({
      type: "action_flow_update",
      flow: mockFlow,
    })
  })

  const [state] = result.current
  expect(state).toEqual({ current: "select_method" })
})

test(`should parse state from flow on "action_flow_update" when choosing method on login`, () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Login,
    flow: {
      active: "",
      ui: {
        nodes: [
          {
            group: "identifier_first",
            attributes: {
              name: "identifier",
              type: "hidden",
            },
          },
        ],
      }, // Assuming nodes structure
    },
  } as unknown as OryFlowContainer

  act(() => {
    dispatch({
      type: "action_flow_update",
      flow: mockFlow,
    })
  })

  const [state] = result.current
  expect(state).toEqual({ current: "select_method" })
})
;[(FlowType.Recovery, FlowType.Verification)].forEach((flowType) => {
  test(`should parse state from flow on "action_flow_update" ${flowType} provide_identifier`, () => {
    const { result } = renderHook(() => useFormStateReducer(init))
    const [, dispatch] = result.current

    const mockFlow = {
      flowType: FlowType.Recovery,
      flow: {
        state: "choose_method",
        active: "code",
        ui: { nodes: [] }, // Assuming nodes structure
      },
    } as unknown as OryFlowContainer

    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    })

    const [state] = result.current
    expect(state).toEqual({ current: "provide_identifier" })
  })

  test(`should parse state from flow on "action_flow_update" ${flowType} flow active`, () => {
    const { result } = renderHook(() => useFormStateReducer(init))
    const [, dispatch] = result.current

    const mockFlow = {
      flowType: FlowType.Recovery,
      flow: {
        state: "sent_email",
        active: "code",
        ui: { nodes: [] }, // Assuming nodes structure
      },
    } as unknown as OryFlowContainer

    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    })

    const [state] = result.current
    expect(state).toEqual({ current: "method_active", method: "code" })
  })
})

test("should throw error for unknown recovery state", () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: FlowType.Recovery,
    flow: { id: "unknown", active: null, ui: { nodes: [] } },
  } as unknown as OryFlowContainer

  expect(() =>
    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    }),
  ).toThrow("Unknown form state")
})

test("should throw error for unrecognized flow", () => {
  const { result } = renderHook(() => useFormStateReducer(init))
  const [, dispatch] = result.current

  const mockFlow = {
    flowType: "unknown" as FlowType, // Intentional to simulate an unexpected flow
    flow: { id: "unknown", active: null, ui: { nodes: [] } },
  } as unknown as OryFlowContainer

  expect(() =>
    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    }),
  ).toThrow("Unknown form state")
})

for (const activeMethod of ["identifier_first", "default"]) {
  test(`should ignore ${activeMethod} group in active field`, () => {
    const { result } = renderHook(() => useFormStateReducer(init))
    const [, dispatch] = result.current

    const mockFlow = {
      flowType: FlowType.Login,
      flow: { id: "unknown", active: activeMethod, ui: { nodes: [] } },
    } as unknown as OryFlowContainer

    act(() => {
      dispatch({
        type: "action_flow_update",
        flow: mockFlow,
      })
    })

    const [state] = result.current
    expect(state).toEqual({ current: "provide_identifier" })
  })
}
