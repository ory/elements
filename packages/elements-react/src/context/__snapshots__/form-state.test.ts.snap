// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse state from flow on "action_flow_update" with active method = code 1`] = `
{
  "current": "method_active",
  "method": "code",
}
`;

exports[`should parse state from flow on "action_flow_update" with active method = code_recovery 1`] = `
{
  "current": "method_active",
  "method": "code",
}
`;

exports[`should parse state from flow on "action_flow_update" with active method = link_recovery 1`] = `
{
  "current": "method_active",
  "method": "link",
}
`;
